src/ui/posts/PostEditor.tsx

```diff
@@ -0,0 +1,91 @@
+import { ReactElement, useEffect, useState } from 'react';
+import { AppBus } from '@/bus/AppBus';
+import { Message } from '@/bus/Message';
+import { ContentBus } from '@/bus/ContentBus';
+import { Post, PostField } from '@/model/content/Post';
+import { FieldEditor } from '@/ui/posts/FieldEditor';
+
+interface Props {
+	post: Post;
+	fieldOrder: Record< string, number >;
+	onFieldChanged: ( name: string, field: PostField ) => void;
+}
+
+export function PostEditor( props: Props ) {
+	const { post, fieldOrder, onFieldChanged } = props;
+	const [ lastClickedElement, setLastClickedElement ] = useState< string >();
+	const [ fieldWaitingForSelection, setFieldWaitingForSelection ] = useState<
+		false | { field: PostField; name: string }
+	>( false );
+
+	// Listen to click events coming from the content script.
+	useEffect( () => {
+		AppBus.listen( async ( message: Message ) => {
+			switch ( message.action ) {
+				case AppBus.actions.ElementClicked:
+					await ContentBus.disableHighlighting();
+					setLastClickedElement( ( message.payload as any ).content );
+			}
+		} );
+		return () => {
+			void ContentBus.disableHighlighting();
+			AppBus.stopListening();
+		};
+	}, [] );
+
+	// Handle a click on an event in the content script,
+	// according to which field is currently waiting for selection.
+	useEffect(
+		() => {
+			if ( ! fieldWaitingForSelection || ! lastClickedElement ) {
+				return;
+			}
+			fieldWaitingForSelection.field.original = lastClickedElement;
+			onFieldChanged(
+				fieldWaitingForSelection.name,
+				fieldWaitingForSelection.field
+			);
+			setFieldWaitingForSelection( false );
+			setLastClickedElement( undefined );
+		},
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ fieldWaitingForSelection, lastClickedElement ]
+	);
+
+	const fields: { name: string; field: PostField }[] = [];
+	for ( const [ name, field ] of Object.entries( post.fields ) ) {
+		const order = fieldOrder[ name ];
+		fields[ order ] = { name, field };
+	}
+
+	const elements: ReactElement[] = [];
+	for ( const { name, field } of fields ) {
+		const isWaitingForSelection =
+			!! fieldWaitingForSelection &&
+			fieldWaitingForSelection.name === name;
+
+		elements.push(
+			<FieldEditor
+				key={ name }
+				label={ name }
+				field={ field }
+				waitingForSelection={ isWaitingForSelection }
+				onWaitingForSelection={ async ( f: PostField | false ) => {
+					await ContentBus.enableHighlighting();
+					if ( !! f ) {
+						setFieldWaitingForSelection( { field: f, name } );
+					} else {
+						setFieldWaitingForSelection( false );
+					}
+				} }
+				onClear={ async () => {
+					field.original = '';
+					field.parsed = '';
+					onFieldChanged( name, field );
```

Comment:
I am a little confused as to why we have to call this manually. Shouldn't we rely on reactivity to take care of such a thing? Or is this just to prevent having to define another prop on `<FieldEditor>`?

Reply:
The component hierarchy here is as follows:

- `EditPost`
    - `PostEditor`
        - `FieldEditor`
        - `FieldEditor`
        - `FieldEditor`
        - ...

The `post` state is owned by `EditPost`, which fetches the post from the API, and puts it into state. The `post` is then passed as a prop to `PostEditor`.

It's also `EditPost` who makes requests to the API to save the post whenever a field changes. Since it's `PostEditor` who knows whether a post field has changed, it needs to tell `EditPost` (the owner of the state) that the state needs to be updated. That's where the `onFieldChanged` event handler comes in. 

So when a field changes, the following happens:

- `PostEditor` calls `onFieldChanged`
- `EditPost` makes a request to the API to save the post
- `EditPost` updates the `post` state
- Components that depend on the `post` state (i.e. components that take `post` as a prop) are rerendered

I'm not sure this answers your question, I hope it makes things clearer.